File: Placer.py
#
# @file   : Placer.py
# @author : Jeb Runnoe
# @date   : Feb 2023
# @brief  : Cell placement optimization infrastructure class
#

import numpy as np
from scipy.fft import idct, idst, dctn, idctn
import scipy.sparse
import matplotlib.pyplot as plt
from matplotlib.patches import Rectangle
from Design import Design
from Net import Net
np.set_printoptions(edgeitems=30, linewidth=100000)

class Placer:
  mu  = 1
  target_density = 0
  smoothing_param = 256
  resolution = 512
  pad_factor = 0.10
  exact = False

  def __init__(self, name):
    self.design = Design(name)
    self.design.extract()

    min_dim = min(self.design.R['dx'], self.design.R['dy'])
    self.grid = dict()
    self.grid['n']   = {'x': int(np.ceil(self.resolution*self.design.R['dx']/min_dim)), 
                        'y': int(np.ceil(self.resolution*self.design.R['dy']/min_dim))}
    self.grid['pad'] = {'x': int(np.ceil(self.pad_factor*self.grid['n']['x'])),
                        'y': int(np.ceil(self.pad_factor*self.grid['n']['y']))}
    self.grid['all'] = {'x': self.grid['n']['x'] + 2*self.grid['pad']['x'],
                        'y': self.grid['n']['y'] + 2*self.grid['pad']['y']}    
    
    self.iofx = lambda x: x + self.grid['pad']['x']
    self.jofy = lambda y: y + self.grid['pad']['y']
    self.xofi = lambda i: i - self.grid['pad']['x']
    self.yofj = lambda j: j - self.grid['pad']['y']

    # Rescale the design so that the grid elements are 1x1
    self.scale = {'x': self.grid['n']['x']/self.design.R['dx'],
                  'y': self.grid['n']['y']/self.design.R['dy']}
    
    self.design.R['x']  *= self.scale['x']
    self.design.R['y']  *= self.scale['y']
    self.design.R['dx'] *= self.scale['x']
    self.design.R['dy'] *= self.scale['y']
    self.design.x0 *= self.scale['x']
    self.design.y0 *= self.scale['y']
    self.design.z0 = np.concatenate((self.design.x0, self.design.y0))
    self.design.dx *= self.scale['x']
    self.design.dy *= self.scale['y']
    self.design.dz = np.concatenate((self.design.dx, self.design.dy))
    self.design.px *= self.scale['x'] 
    self.design.py *= self.scale['y']
    self.design.pz = np.concatenate((self.design.px, self.design.py))

    # convenience definitions
    self.ix = range(self.design.n_cells)
    self.iy = range(self.design.n_cells, 2*self.design.n_cells)
    self.half_dx = np.divide(self.design.dx, 2)
    self.half_dy = np.divide(self.design.dy, 2)
    self.half_dz = np.divide(self.design.dz, 2)

    self.nets = np.ndarray(self.design.n_nets, dtype=object)
    for j in range(self.design.n_nets):
      self.nets[j] = Net(j, self.design.netlist[j], self.design.px, self.design.py)

    # bound constraints
    self.lx = self.half_dx
    self.ly = self.half_dy
    self.lz = self.half_dz
    self.ux = self.design.R['dx']*np.ones_like(self.half_dx) - self.half_dx
    self.uy = self.design.R['dy']*np.ones_like(self.half_dy) - self.half_dy
    self.uz = np.concatenate((self.ux, self.uy))

    # electrostatic quatities: 
    (self.density, self.masks) = self.compute_density(self.design.x0, self.design.y0) # charge density map
    self.potential = np.empty((self.grid['all']['x'], self.grid['all']['y']))         # electrostatic potential map
    self.field     = {'x': np.empty((self.grid['all']['x'], self.grid['all']['y'])),  # electrostatic x & y component field maps
                      'y': np.empty((self.grid['all']['x'], self.grid['all']['y']))} 
            
  def evaluate(self, z):
    x = np.median((self.lx - self.grid['pad']['x'], z[self.ix], self.ux + self.grid['pad']['x']), axis=0)
    y = np.median((self.ly - self.grid['pad']['y'], z[self.iy], self.uy + self.grid['pad']['y']), axis=0)

    (D, M)    = self.compute_density(x, y)
    [P, F, E] = self.compute_electrostatics(D)

    # Determine the grid element(s) in which each cell locates, extract the
    # electric field at those locations, and scale by the cell charges (area). 
    dE = {'x': np.zeros_like(x), 'y': np.zeros_like(y)}

    for k in range(self.design.n_cells):
      [I, J] = self.masks[k].nonzero()
      if self.exact:
        dE['x'][k] = -np.multiply(F['x'][I, J], M[k][I, J]).sum()
        dE['y'][k] = -np.multiply(F['y'][I, J], M[k][I, J]).sum()
      else:
        dE['x'][k] = -F['x'][I, J].sum()
        dE['y'][k] = -F['y'][I, J].sum()

    W    = 0.0
    dW = {'x': np.zeros_like(x), 'y': np.zeros_like(y)}

    for j in range(self.design.n_nets):
      (nW, ndW) = self.nets[j].pin_variance(x, y)
      W += nW
      dW['x'][self.nets[j].cells] += ndW['x']
      dW['y'][self.nets[j].cells] += ndW['y']

    df = {'x': dW['x'] + self.mu*dE['x'], 
          'y': dW['y'] + self.mu*dE['y']}
    f  =  W + self.mu*E
    g  = np.concatenate((df['x'], df['y']))
    return(f, g)

  def update_density(self, x, y, cells, D, M):
    '''
    @brief: Update the density map by subtracting, shifting, and adding density masks of provided cells
    '''
    for k in cells:
      di = int(np.floor(self.iofx(x[k] - self.half_dx[k]))) - M[k].nonzero()[0][0] # Change in row index
      dj = int(np.floor(self.jofy(y[k] - self.half_dy[k]))) - M[k].nonzero()[1][0] # Change in column index

      if di != 0 or dj != 0: 
        D = D._add_sparse(-M[k]) # Subtract out this cells previous density contribution

        indices = M[k].indices + dj # Shift column indices 
    
        # Shift row index pointers (in CSR format)
        # For example,
        # shift   up 3 rows:  [0, 0, 0, 0, 4, 9, 13, 18, 18, 18, 18, 18] -> [0, 4, 9, 13, 18, 18, 18, 18, 18, 18, 18, 18]
        # shift down 2 rows:  [0, 0, 0, 0, 4, 9, 13, 18, 18, 18, 18, 18] -> [0, 0, 0,  0,  0,  0,  4,  9, 13, 18, 18, 18]
        if di > 0:
          pointers = np.concatenate((np.zeros(di), M[k].indptr[:-di])).astype(int)
        elif di < 0: 
          pointers = np.concatenate((M[k].indptr[abs(di):], M[k].nnz*np.ones(abs(di)))).astype(int)
        else: 
          pointers = M[k].indptr
        
        M[k] = scipy.sparse.csr_matrix((M[k].data, indices, pointers), shape=(self.grid['all']['x'], self.grid['all']['y']))

        D = D._add_sparse(M[k])
    D.data = np.maximum(D.data - self.target_density, 0)
    return (D, M)

  def compute_density(self, x, y):
    '''
    @brief   : computation of the cell density map based on the current placement
    @param x : current placement x-coordinates
    @param y : current placement y-coordinates
    '''
    D = scipy.sparse.csr_matrix((self.grid['all']['x'], self.grid['all']['y']))
    M = [None]*self.design.n_cells

    for k in range(self.design.n_cells):
      lx = int(np.floor(x[k] - self.half_dx[k]))
      ly = int(np.floor(y[k] - self.half_dy[k]))
      ux = int(np.floor(x[k] + self.half_dx[k]))
      uy = int(np.floor(y[k] + self.half_dy[k]))
      dx_k = ux - lx
      dy_k = uy - ly

      if self.exact:
        block = np.zeros((dx_k + 1, dy_k + 1))      

        # Width and height of partial overlaps
        ldx = min((lx + 1) - (x[k] - self.half_dx[k]), self.design.dx[k]) # lower horizontal
        ldy = min((ly + 1) - (y[k] - self.half_dy[k]), self.design.dy[k]) # lower vertical
        udx = (x[k] + self.half_dx[k]) - ux if dx_k > 0 else 0 # upper horizontal
        udy = (y[k] + self.half_dy[k]) - uy if dy_k > 0 else 0 # upper vertical

        block[0,           0] += ldx*ldy # lower left 
        block[0,        dy_k] += ldx*udy # upper left 
        block[  dx_k,      0] += udx*ldy # lower right
        block[  dx_k,   dy_k] += udx*udy # upper right
        block[1:dx_k,      0] += ldy     # lower horizontal strip
        block[0,      1:dy_k] += ldx     # left vertical strip
        block[  dx_k, 1:dy_k] += udx     # right vertical strip
        block[1:dx_k,   dy_k] += udy     # upper horizontal strip
        block[1:dx_k, 1:dy_k] += 1.0     # interior
      else:
        block = np.ones((dx_k + 1, dy_k + 1))

      [I, J] = np.meshgrid(self.iofx(np.arange(lx, ux + 1)), self.jofy(np.arange(ly, uy + 1)), indexing='ij')
      M[k]   = scipy.sparse.csr_matrix((block.ravel(), (I.ravel(), J.ravel())), shape=(self.grid['all']['x'], self.grid['all']['y']))
      D      = D._add_sparse(M[k])
    D.data = np.maximum(D.data - self.target_density, 0)
    return(D, M)

  def compute_electrostatics(self, D):
    '''
    @brief : Computation of the system electrostatic quantities determined by the density map
    @param : D is a matrix with the density value at each grid element 
    '''
    if scipy.sparse.issparse(D):
      D = D.todense() # DFT routines don't work with sparse arrays
      
    (M, N) = D.shape

    # Define the frequency mesh for the cosine bases
    u = (np.pi/M)*np.arange(M)
    v = (np.pi/N)*np.arange(N)

    # compute the Fourier coefficients to express the density map D in the cosine basis 
    # over the u-v frequency mesh.
    A = dctn(D, type=2, norm="ortho")

    # the electrostatic potential map is obtained by integrating the cosine representation
    # of the density map over the placement region
    S       = np.add.outer(np.square(u), np.square(v)) # (S)um of squares matrix
    S[0, 0] = 1
    C       = np.divide(A, S)
    C[0, 0] = 0
    P       = idctn(C, type=2, norm="ortho")

    F = dict.fromkeys(['x', 'y']) 
    # Differentiating the potential map with respect to x gives the 
    # horizontal component of the electrostatic field. This essentially multiplies
    # each row of the coefficient matrix C by elements of u
    Cu = np.multiply(C.T, u).T    
    F['x'] = idst(np.vstack((idct(Cu, axis=1, norm="ortho")[1:, :], np.zeros([1, N]))), axis=0, norm="ortho")
     
    # Differentiating the potential map with respect to y gives the vertical
    # component of the electrostatic field. This is done by multiplying the
    # indices of C by elements of v.
    Cv = np.multiply(C, v)
    F['y'] = idct(idst(np.hstack((Cv[:, 1:], np.zeros([M, 1]))), axis=1, norm="ortho"), axis=0, norm="ortho")

    E = 0.50*np.multiply(P, D).sum()    
    return(P, F, E)

  def plot_cells(self, x, y, cells="all"):
    plt.xlim(-1.25*self.grid['pad']['x'], self.grid['n']['x'] + 1.25*self.grid['pad']['x'])
    plt.ylim(-1.25*self.grid['pad']['y'], self.grid['n']['y'] + 1.25*self.grid['pad']['y'])
    plt.axhline(0, color='black', linewidth=.5)
    plt.axvline(0, color='black', linewidth=.5)

    if cells == "all":
      cells = range(self.design.n_cells)

    # Show the entire padded region
    plt.gca().add_patch(Rectangle((-self.grid['pad']['x'], -self.grid['pad']['y']), 
                                    self.grid['n']['x'] + 2*self.grid['pad']['x'],
                                    self.grid['n']['y'] + 2*self.grid['pad']['y'],
                                    edgecolor='red', facecolor='none', alpha=1, lw=1))
    
    # Show the actual placement region
    plt.gca().add_patch(Rectangle((0, 0), self.design.R['dx'], self.design.R['dy'], 
                                   edgecolor='green', facecolor=[0, 1, 0, .05], ls='--'))

    for i in cells:
      plt.gca().add_patch(Rectangle((x[i] - self.half_dx[i], y[i] - self.half_dy[i]), 
                                     self.design.dx[i], self.design.dy[i], 
                                     edgecolor='black', facecolor='none', alpha=1))

  def plot_heatmap(self, Z):
    plt.xlim(-1.25*self.grid['pad']['x'], self.grid['n']['x'] + 1.25*self.grid['pad']['x'])
    plt.ylim(-1.25*self.grid['pad']['y'], self.grid['n']['y'] + 1.25*self.grid['pad']['y'])
    plt.axhline(0, color='black', linewidth=.5)
    plt.axvline(0, color='black', linewidth=.5)
    X, Y = np.meshgrid(self.xofi(np.arange(Z.shape[0])), self.yofj(np.arange(Z.shape[1])), indexing='ij')
    hm = plt.gca().pcolormesh(X, Y, Z)
    plt.colorbar(hm)

  def plot_contour(self, Z):
    plt.xlim(-1.25*self.grid['pad']['x'], self.grid['n']['x'] + 1.25*self.grid['pad']['x'])
    plt.ylim(-1.25*self.grid['pad']['y'], self.grid['n']['y'] + 1.25*self.grid['pad']['y'])
    plt.axhline(0, color='black', linewidth=.5)
    plt.axvline(0, color='black', linewidth=.5)
    X, Y = np.meshgrid(self.xofi(np.arange(Z.shape[0])), self.yofj(np.arange(Z.shape[1])), indexing='ij')
    plt.gca().contour(X, Y, Z, 30)

  def plot_surface(self, Z):
    X, Y = np.meshgrid(self.xofi(np.arange(Z.shape[0])), self.yofj(np.arange(Z.shape[1])), indexing='ij')
    plt.gca(projection='3d').plot_surface(X, Y, Z)

  @staticmethod
  def idsxt(X, axis=1):
    if axis == 0:
      X = X.T
    [m, n] = X.shape
    Y = np.zeros_like(X)
    for i in range(m):
      y = idct(np.flip(np.append(X[i, 1:], 0)), norm="ortho")
      y[1::2] = -y[1::2]
      Y[i,:]  = y
    if axis == 0:
      Y = Y.T
    return Y

  @staticmethod
  def overlap(r1, r2):
    ''' 
    Computes the area of the intersection of two rectangles r1 and r2
    '''
    dx = max(0, min(r1['x'] + 0.50*r1['dx'], r2['x'] + 0.50*r2['dx']) - max(r1['x'] - 0.50*r1['dx'], r2['x'] - 0.50*r2['dx']))
    dy = max(0, min(r1['y'] + 0.50*r1['dy'], r2['y'] + 0.50*r2['dy']) - max(r1['y'] - 0.50*r1['dy'], r2['y'] - 0.50*r2['dy']))
    return(dx*dy)

File: Placer.m
classdef Placer < handle
  properties
    design;
    chip;
    nets;
    grid;
    masks;
    scale;

    % Electrostatic quantities
    density;
    potential;
    field;
    
    ix;       % indices of x coordinates (  1,   2, ...,   n)
    iy;       % indices of y coordinates (n+1, n+2, ..., 2*n)
    half_dx;
    half_dy;
    half_dz;

    uz;
    lz;

    mu    = 0.01; % Multiplier for energy 
    sigma = 1;    % Multiplier for bound constraint violation
  end
  properties(Constant)
    grid_resolution = 128;
    pad_factor = 0.50;
    step_contraction = 0.67;
    reduction_tolerance = 1.0e-04;
    df_tolerance = 1.0e-01;
    colors = colororder;
  end
  
  methods
    function [obj] = Placer()
      obj.design = generate_design();
      
      obj.nets = cell(obj.design.n_nets, 1);
      for j = 1:obj.design.n_nets
        obj.nets{j} = Net(j, obj.design.netlist{j});
      end
      
      obj.grid.nx = ceil(obj.grid_resolution*obj.design.R.dx/min(obj.design.R.dx, obj.design.R.dy));
      obj.grid.ny = ceil(obj.grid_resolution*obj.design.R.dy/min(obj.design.R.dx, obj.design.R.dy));
      obj.grid.pad.x = ceil(obj.pad_factor*obj.grid.nx);
      obj.grid.pad.y = ceil(obj.pad_factor*obj.grid.ny);
      obj.grid.pad.z = [obj.grid.pad.x*ones(obj.design.n_cells, 1); obj.grid.pad.y*ones(obj.design.n_cells,1)];
      obj.grid.total.x = obj.grid.nx + 2*obj.grid.pad.x;
      obj.grid.total.y = obj.grid.ny + 2*obj.grid.pad.y;

      obj.scale.x = obj.grid.nx/obj.design.R.dx;
      obj.scale.y = obj.grid.ny/obj.design.R.dy;

      obj.chip.R  = struct('x', 0, 'y', 0, 'dx', obj.grid.nx, 'dy', obj.grid.ny);
      obj.chip.x0 = obj.scale.x*obj.design.x0;
      obj.chip.y0 = obj.scale.y*obj.design.y0;
      obj.chip.z0 = [obj.chip.x0; obj.chip.y0];
      obj.chip.dx = obj.scale.x*obj.design.dx;
      obj.chip.dy = obj.scale.y*obj.design.dy;
      obj.chip.dz = [obj.chip.dx; obj.chip.dy];
       
      obj.ix = 1:obj.design.n_cells;
      obj.iy = (obj.design.n_cells + 1):2*obj.design.n_cells;
      obj.half_dx = 0.50*obj.chip.dx;
      obj.half_dy = 0.50*obj.chip.dy;
      obj.half_dz = [obj.half_dx; obj.half_dy];

      obj.lz = obj.half_dz;
      obj.uz = [obj.chip.R.dx*ones(obj.design.n_cells, 1); 
                obj.chip.R.dy*ones(obj.design.n_cells, 1)] - obj.half_dz;

      [obj.density, obj.masks] = obj.compute_density(obj.chip.x0, obj.chip.y0);

      obj.potential = zeros(obj.grid.nx, obj.grid.ny);
      obj.field.x   = zeros(obj.grid.nx, obj.grid.ny);
      obj.field.y   = zeros(obj.grid.nx, obj.grid.ny);
    end

    function [] = plot_cells(obj, x, y, cells)
      if isempty(cells)
        cells = 1:obj.design.n_cells;
      end
      rectangle('Position', [obj.chip.R.x,  obj.chip.R.y,   ...
                             obj.chip.R.dx, obj.chip.R.dy], ...
                             'EdgeColor', obj.colors(1, :), 'FaceColor', [obj.colors(1,:) 0.05], "LineStyle","--");
      hold on;
      for i = cells
        color = 'black';
        for j = 1:7
          if ismember(i, obj.nets{j}.cells)
            color = obj.colors(j, :);
            break;
          end  
        end
        rectangle('Position', [x(i) - obj.half_dx(i), ...
                               y(i) - obj.half_dy(i), ...
                               obj.chip.dx(i), ...
                               obj.chip.dy(i)], 'EdgeColor', color);
      end
      axis([-obj.grid.pad.x, obj.chip.R.dx + obj.grid.pad.x, -obj.grid.pad.y, obj.chip.R.dy + obj.grid.pad.y]);
      xline(0, "LineStyle",":");
      yline(0, "LineStyle",":");

      hold off;
    end

    function [hm] = plot_heatmap(obj, M)
      hm = imagesc(M');
      set(gca, 'Ydir', 'normal');
      hm.XData = hm.XData - obj.grid.pad.x - 1;
      hm.YData = hm.YData - obj.grid.pad.y - 1;
      colormap('parula');
      colorbar;
      rectangle('Position', [obj.chip.R.x,  obj.chip.R.y,   ...
                             obj.chip.R.dx, obj.chip.R.dy], ...
                             'EdgeColor', obj.colors(1, :), "LineStyle","--");
      
      axis([-obj.grid.pad.x, obj.chip.R.dx + obj.grid.pad.x - 1, -obj.grid.pad.y, obj.chip.R.dy + obj.grid.pad.y - 1]);
    end

    function [clearance] = compute_clearance(obj, x, y, cells)
      %% ----------------- compute_clearance()
      %  Computation of the minimum pairwise clearance for the provided
      %  group of cells. The result is positive if there is no overlap,
      %  zero when two or more cells are touching, and negative if two or
      %  more cells intersect.
      
      pairs   = nchoosek(cells, 2);
      n_pairs = size(pairs, 1);
      deltas  = zeros(n_pairs, 1);

      for k = 1:n_pairs
        xi = x(pairs(k, 1)); 
        xj = x(pairs(k, 2));
        yi = y(pairs(k, 1)); 
        yj = y(pairs(k, 2));

        dxi = obj.chip.dx(pairs(k, 1));
        dxj = obj.chip.dx(pairs(k, 2));
        dyi = obj.chip.dy(pairs(k, 1));
        dyj = obj.chip.dy(pairs(k, 2));
        
        dijx = max(xi - 0.50*dxi, xj - 0.50*dxj) - min(xi + 0.50*dxi, xj + 0.50*dxj);
        dijy = max(yi - 0.50*dyi, yj - 0.50*dyj) - min(yi + 0.50*dyi, yj + 0.50*dyj);

        deltas(k) = max(dijx, dijy);
      end
      clearance = min(deltas);
    end

    function [result] = solve(obj)
      z = obj.chip.z0;

      converged = false;
      [f, g, ~, ~] = obj.evaluate(z);
      
      while ~converged
        p = -g;
        [z1, f1, g1, D, M] = obj.search(z, p, f, g);

        df = f1-f
        
        if abs(df) < obj.df_tolerance
          converged = true;
        end

        % Set the new base point
        obj.density = D;
        obj.masks = M;
        z = z1;
        f = f1;
        g = g1;
        
      end
      result.x = z(obj.ix);
      result.y = z(obj.iy);
      result.z = z;
      result.f = f;
    end

    function [z1, f1, g1, D, M] = search(obj, z, p, f, g)
      
      reduced = false;
      alpha = 1;
      gTp = g'*p;

      while ~reduced && alpha > eps
        z1 = z + alpha*p;

        z1 = median([obj.lz - obj.grid.pad.z + 1, z1, obj.uz + obj.grid.pad.z - 1], 2);

        [f1, g1, D, M] = obj.evaluate(z1);
        
        if f - f1 >= -obj.reduction_tolerance*alpha*gTp
          reduced = true;
        else
          alpha = obj.step_contraction*alpha;
        end
      end
    end

    function [f, g, D, M] = evaluate(obj, z)
      %% ------------ evaluate()
      %
      %  DESCRIPTION
      %  This computes the objective function 
      %       f(z) = W(z) + mu*E(z)
      %  and its gradient, where W is some measure of the
      %  wire length and E is the system electrostatic energy.
      %  Possible choices for W(z) include the weighted average
      %  smooth approximation of half-perimeter wire length, or
      %  the pin coordinate variances.
      %  
      %  Parameters
      %  z: the current placement coordinates z = [x; y].
      %  D: the previous density matrix  - will be updated to reflect the new placement z.
      %  M: the previous occupancy masks - will be updated to reflect the new placement z.
      % 
      %  Returns
      %  f: the function value.
      %  g: the gradient of f at z.
      %  D: updated density matrix.
      %  M: updated occupancy masks.
      
      x = z(obj.ix); 
      y = z(obj.iy);

      [D, M]    = obj.compute_density(x, y);

      P   = sparse(obj.grid.total.x, obj.grid.total.y);
      F.x = sparse(obj.grid.total.x, obj.grid.total.y);
      F.y = sparse(obj.grid.total.x, obj.grid.total.y);
      
      % If any cells extend out of the main (grid.nx by grid.ny)
      % grid, extend the grid into the pad to capture all cells.
      xmin = min([1; floor(x - obj.half_dx) + 1]);
      xmax = max([obj.grid.nx; ceil(x + obj.half_dx)]);
      ymin = min([1; floor(y - obj.half_dy) + 1]);
      ymax = max([obj.grid.ny; ceil(y + obj.half_dy)]);
      I = obj.grid.pad.x + (xmin:xmax);
      J = obj.grid.pad.y + (ymin:ymax);
      
      [~, F_, E] = obj.compute_electrostatics(D(I, J));

      F.x(I, J) = F_.x;
      F.y(I, J) = F_.y;

      % Determine in which grid elements each cell locates, extract 
      % the electric field at those locations, and scale by the cell 
      % charges (area):
      dE.x = zeros(size(x));
      dE.y = zeros(size(y));

      for k = 1:obj.design.n_cells
        [I, J] = find(M{k});
        % The total force on cell k will be the sum of the forces
        % due to the electrostatic field at each grid element scaled by the overlap
        % area there.
        dE.x(k) = -sum(F.x(sub2ind(size(F.x), I, J)), "all");
        dE.y(k) = -sum(F.y(sub2ind(size(F.y), I, J)), "all");
      end

      % The wirelength gradient is the sum over nets 
      dW.x = zeros(size(x));
      dW.y = zeros(size(y));
      W = 0;
      for j = 1:obj.design.n_nets
        [nW, ndW] = obj.nets{j}.pin_spread(x, y);
        % [nW, ndW] = obj.nets{j}.WAWL(x, y);
        W = W + nW;
        dW.x(obj.nets{j}.cells) = dW.x(obj.nets{j}.cells) + ndW.x;
        dW.y(obj.nets{j}.cells) = dW.y(obj.nets{j}.cells) + ndW.y;
      end
      
      % Add the norm of the bound constraint residuals to the
      % objective and its gradient
      
      upr_resid = max(z - obj.uz, 0);
      low_resid = max(obj.lz - z, 0);

      R = 0.50*norm( upr_resid )^2 + 0.50*norm( low_resid )^2;
      dR = upr_resid - low_resid;

      f = W + obj.mu*E + obj.sigma*R;
      g = [dW.x + obj.mu*dE.x;
           dW.y + obj.mu*dE.y] + obj.sigma*dR;
    end

    function [D, M] = update_density(obj, x, y, cells)
      %% ------------ update_density()
      %
      %  Description
      %  Updates the density matrix D and occupancy masks M based
      %  on new coordinates x, y. It is assumed only cells listed
      %  change position. The update is done by subtracting out
      %  the density contribution of the moved cells, updating
      %  their sparse density mask, and adding the new
      %  contribution back into D. 
      %
      %  Parameters
      %  x    : new cell coordinates
      %  y    : new cell coordinates
      %  D    : (sparse double) current density matrix.
      %  M    : (cell array) cell array of sparse occupancy masks.
      %  cells: list of cells to update.
      %
      %  Returns
      %  D: (sparse double) Density matrix
      %  M: (cell array) Grid-cell occupancy masks
      %
      D = obj.density;
      M = obj.masks;
      for k = cells
        [I, J, v] = find(obj.masks{k});
        di = floor(x(k) - obj.half_dx(k)) - I(1) + 1;
        dj = floor(y(k) - obj.half_dy(k)) - J(1) + 1; 
        if di ~= 0 || dj ~= 0
          D    = D - obj.masks{k};
          M{k} = sparse(I + di, J + dj, v, obj.grid.nx, obj.grid.ny);
          D    = D + M{k};
        end
      end
    end

    function [D, M] = compute_density(obj, x, y)
      %% ------------ compute_density()
      %
      %  Description
      %  Compute the density based on the placement coordinates
      %  x, y. The instances listed in the cells parameter are
      %  assumed to have moved relative to the coordinates used
      %  to compute the provided D, M. If cells lists all design
      %  cells, the density is computed from scratch, otherwise
      %  the density is updated.
      %
      %  Parameters
      %  x    : new cell coordinates
      %  y    : new cell coordinates
      %  D    : (sparse double) current density matrix.
      %  M    : (cell array) cell array of sparse occupancy masks.
      %  cells: list of cells to update.
      %
      %  Returns
      %  D: (sparse double) Density matrix
      %  M: (cell array) Grid-cell occupancy masks
      %

      D = sparse(obj.grid.total.x, obj.grid.total.y);
      M = cell(obj.design.n_cells, 1);
      for k = 1:obj.design.n_cells
        li = floor(x(k) - obj.half_dx(k)) + 1;
        ui = ceil(x(k)  + obj.half_dx(k));
        lj = floor(y(k) - obj.half_dy(k)) + 1;
        uj = ceil(y(k)  + obj.half_dy(k));
        
        M{k} = sparse(obj.grid.total.x, obj.grid.total.y);
        M{k}(obj.grid.pad.x + (li:ui), obj.grid.pad.y + (lj:uj)) = 1;
        % The density is the sum over individual grid-cell occupancy masks
        D = D + M{k};
      end
    end

    function [P, F, E] = compute_electrostatics(obj, D)
      %% --------------- compute_electrostatics()
      %
      %  Description
      %  Computes the electrostatic potential map, horizontal and
      %  vertical field maps, and total energy using fast Fourier
      %  transforms
      %
      %  Parameters
      %  D: density matrix
      %
      %  Returns
      %  P: (double) potential map such that P, D satisfy Poisson's equation
      %  F: (struct) electric field maps F.x and F.y
      %  E: (scalar) total system energy

      if issparse(D)
        D = full(D);
      end
      [M, N] = size(D);
      % Generate the mesh of support for the cosine basis
      u  = (pi/M)*(0:(M - 1));
      v  = (pi/N)*(0:(N - 1));

      % Compute the Fourier coefficients to express the density D in the
      % cosine basis. The 2D DCT is achieved by iterating 1D DCT
      A = dct(dct(D, [], dim=1), [], dim=2);

      % Compute the electrostatic potential: (the coefficients
      % A./S come from rho(x, y) = -laplacian(psi(x, y)), writing the expansion
      % of the density in the cosine basis, and integrating twice
      % in each variable.
      S       = (u.^2)' + v.^2; % (S)um of squares
      S(1, 1) = 1;
      C       = A./S;
      C(1, 1) = 0;

      P = idct(idct(C, [], dim=1), [], dim=2);

      % The force calculation can also be done using the idst function:
      % T   = idct(Cu')';
      % F.x = idst([T(2:end, :); zeros(1, obj.grid.ny)]);
      % F.y = idct(idst([Cv(:, 2:end) zeros(obj.grid.nx, 1)]')');

      % Differentiating the potential map with respect to x gives the 
      % horizontal component of the electrostatic field. This multiplies
      % each row of the coefficient matrix C by elements of u.
      Cu  = u'.*C;
      F.x = obj.idxst(idct(Cu, [], "dim", 2));

      % Differentiating the potential map with respect to y gives the
      % vertical component of the electrostatic field.
      Cv  = C.*v;
      F.y = idct(obj.idxst(Cv, [], "dim", 2));

      % Total system energy is half the sum over all charge-potential
      % products (charge represented by area)
      E = 0.50*sum(P.*D, "all");
    end
  end

  methods(Static)
    function [area] = compute_overlap(r1, r2)
      %% ------------ compute_overlap(r1, r2)
      % Compute the area of the overlap of two rectangles r1, r2. 
      dx = max(0, min(r1.x + r1.dx, r2.x + r2.dx) - max(r1.x, r2.x));
      dy = max(0, min(r1.y + r1.dy, r2.y + r2.dy) - max(r1.y, r2.y));
      area = dx*dy;
    end

    function [Y] = idxst(X, ~, varargin)
      %% --------- idxst(X, ~, varargin)
      % Purpose: 
      % Compute the inverse discrete sine transform of the input
      % transformed by [a_1, a_2, a_3,..., a_n] -> [a_2, a_3,..., a_n, 0]
      % in terms of the idct function (by reversing the input and flipping 
      % the sign of every other output).
      if nargin > 2
        dim = varargin{2};
      else
        dim = 1;
      end
      if dim == 1
        X = X';
      end
      [m, ~] = size(X);
      Y = zeros(size(X));
      for i = 1:m
        y = idct(flip([X(i, 2:end) 0]));
        y(2:2:end) = -y(2:2:end);
        Y(i, :) = y;
      end
      if dim == 1
        Y = Y';
      end
    end
  end
end